### ISSUE TITLE
Possible bug when undoing last move, and useRef question with Typescript

### ISSUE THREAD

**Issue by amcarthur** [Clariity/react-chessboard#5]
Following the example here: https://github.com/Clariity/react-chessboard/blob/main/example/src/boards/PlayVsPlay.js

It seems as though we need to call `.clearPremoves()` after undoing a move.

First, this seems like a bug or anti-pattern. Right now, calling `.undo()` without calling `.clearPremoves()` seems to not update the last piece color, resulting in the color that is supposed to be able to move not being able to ([this](https://github.com/Clariity/react-chessboard/blob/a896ef85a8335eb9ebb81d375627f1bb3f6b8951/src/context/chessboard-context.js#L184) if statement evaluates to `true` because `lastPieceColour === piece[0]`). When you attempt to move a piece of the other color, it still won't work, but some branching side effect of that check (I think [here](https://github.com/Clariity/react-chessboard/blob/a896ef85a8335eb9ebb81d375627f1bb3f6b8951/src/context/chessboard-context.js#L215) because my onPieceDrop handler correctly returns `false` because the underlying call to `game.move()` fails) corrects the color that is allowed to move, seemingly by calling `clearPremoves()`. If this is intended, the function should probably be renamed, or the last color logic refactored into another function, because the intention is unclear from that name. Or better yet, encapsulate this so we don't need to call any function after calling `.undo()`.

Second, to call `.clearPremoves()` you need to use a ref (I think). However, Typescript is complaining that there is no ref property on Chessboard:

```
Type '{ id: number; position: string; onPieceDrop: (sourceSquare: Square, targetSquare: Square) => boolean; onSquareClick: (square: Square) => void; customBoardStyle: { ...; }; customSquareStyles: {}; boardWidth: number; ref: RefObject<...>; }' is not assignable to type 'IntrinsicAttributes & ChessBoardProps'.
  Property 'ref' does not exist on type 'IntrinsicAttributes & ChessBoardProps'.  TS2322

    176 |               }}
    177 |               boardWidth={boardWidth}
  > 178 |               ref={chessboardRef}
        |               ^
    179 |             />
```

My useRef:
`const chessboardRef = useRef();`

My Chessboard, where the error is originating from:
```
<Chessboard
  ...
  ref={chessboardRef}
/>
```

This might be just be a lack of understanding on my part with Typescript/React/useRef. If so, any help would be appreciated. Regardless of this ref issue, I think the first issue I pointed out is still a bug.

Thanks

**Comment by amcarthur** [comment]
I'm using Node 16.13.0, React 17.0.2, Typescript 4.5.2, and react-chessboard 0.0.16

**Comment by amcarthur** [comment]
Setting `expectingAlternateMoves` seems to fix this because the chess.js engine handles that logic. If it fails to `.move()`, I return false in the `onPieceDrop` handler. This seems like a hacky workaround though.

**Comment by Clariity** [comment]
Hi @amcarthur, thanks for raising this issue, I'll give it a proper look after I finish work

- For the useRef bit, I assume that's just because its missing from the type declaration file
- For the clearing premoves, ill give it a good look later and see if it can be incorporated automatically without having to explicitly clear them after an undo

**Comment by Clariity** [comment]
@amcarthur 

Upon further investigation, it seems as though you were referencing code within the `handleSetPosition` function. This function is only called upon a manual drag and drop, and as such, wouldn't be called in the event of an external position change by undo().

Game logic should be handled outside of the react-chessboard component. The board has no idea if it needs to clear pre-moves or not based on an external position change, as that external position change could very well be a valid next move.

To put it more simply, undo() isn't a feature that's implemented by react-chessboard, so undoing a position (and as a result clearing any pre-moves) should be handled externally to the package, but because I implemented pre-move capability to the package, I also exposed the ability to clear them if needed.

I will release 0.0.17 now with the updated type declaration file for ref, if you could upgrade and confirm the error message has disappeared that would be helpful. And thank you again for bringing this issue up.

**Comment by amcarthur** [comment]
@Clariity Thanks so much for looking into this!

What you said about the chessboard component having no knowledge of external position changes makes a lot of sense.

I still think that having to call `.clearPremoves()` after an external state-change occurs is not super intuitive because, for example, I have pre-moves disabled. So I didn't think I needed to call a function with "premove" in the name until I looked at your examples.

Also, it looks like you updated src/types/index.d.ts but not dist/index.d.ts

Thanks for your help, I'll close this since I think everything will work fine after you push the dist updates.

**Comment by Clariity** [comment]
@amcarthur 

I don't believe clearPremoves() is required to be called if premoves are disabled. Its an optional function to call if you have premoves enabled and you want to clear them

Cheers for pointing that out, I'll update the dist and re-release

**Comment by amcarthur** [comment]
@Clariity Maybe this is a bug then. Here's a screen recording with premoves disabled and without calling clearPremoves() that shows what happens after I .undo() a move. https://www.youtube.com/watch?v=ZBz14BRTFDw

**Comment by Clariity** [comment]
Yep you're right, that's a bug, I'll start looking into it, thank you for the helpful recording, makes it easy to reproduce

**Comment by Clariity** [comment]
@amcarthur Hopefully fixed and released in new version 1.0.0 (will likely require reinstall as is a major version change)

**Comment by amcarthur** [comment]
Fixed! Thank you.
